accepted/2025/PrunePackageReference-with-direct-PackageReference.md

```diff
@@ -0,0 +1,264 @@
+With a pruned reference, the package is explicitly not available, but not in an error state.
+In addition, a pruned package should not be audited for vulnerabilities, but ideally is audited for deprecations.
+
+For the components mentioned above, we need to define their behavior for pruned packages.
+
+| Component | Pruned direct PackageReference behavior | Notes |
+|-----------|-----------------------------------------|-------|
+| Package Manager UI (Installed tab) | Show | The customer needs to see that they've specified the PackageReference and should be able to remove that package. Only the requested version will be used for any logic, that means requested/resolved are expected to the same. |
+| Package Manager UI (Updates) | Show | An update to the package may lead to it not being pruned anymore, and the customer needs to be able to make that decision. Only the requested version will be used for any logic, that means requested/resolved are expected to the same. |
+| dotnet list package | Show | Similar as Package Manager UI, installed tab. The list package has a requested/resolved column. The resolved column should match the requested, as well as an additional indicator similar to auto-referenced. |
+| dotnet list package --outdated | Show | A customer needs to be able to update, as an update may make a package not pruned anymore. The resolved column should match the requested, as well as an additional indicator similar to auto-referenced. |
+| dotnet nuget why | Show | Similar reasoning as list package. |
+| Solution Explorer | Show | The package needs to indicate why it was not "resolved". The solution explorer currently adds a warning icon next to unresolved packages (ie, not found in the project.assets.json file). We should change the indicator to something different. |
+| GetInstalledPackagesAsync | Show | Depends on the scenario, they may be trying to achieve similar behavior to what the PM UI and list package commands provide. The [NuGetInstalledPackage](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.VisualStudio.Contracts/NuGetInstalledPackage.cs) type from the INuGetProjectService (that test explorer and roslyn use), will have the following changes: `Version` => null, the documentation will be updated to say the package is in either an error case or pruned, InstallPath => null, similarly if the package does not exist, the installation path does not exist. |
+
+Given that pruning is an offline operation, ie NuGet does not talk to the sources for pruned packages, we never validate that the requested package exists and the specified version may be incorrect. This is likely to not be of too much consequence, since it's only really possible with hand-editing, and the impact of it is minimal since the package is not used anyways.
+
+## Prior Art
+
+N/A
+
+## Unresolved Questions
+
+- How should we handle direct packages? `PrivateAssets=all`, prune completely or no action.
```

Comment:
I've come around on not pruning directs.  I think treating as private is "better" with some sort of escape hatch.  I don't think it should be a hardcoded TFM version, but instead some property that folks could toggle if they wish.

Reply:
I was originally not leaning towards allowing a switch for how directs are handled, kind of a this is how pruning works thing. 

I agree with regards to the TFM. I'll need to send details about this to the breaking change aliases.